src/box/lua/tuple.c

```diff
@@ -705,14 +733,205 @@ static const struct luaL_Reg lbox_tuple_meta[] = {
+}
+
+/*
+ * Pushes format of tuple field onto Lua stack.
+ */
+static void
+push_tuple_field_format(struct lua_State *L, struct tuple_format *fmt,
+			uint32_t field_no)
+{
+	struct tuple_field *field = tuple_format_field(fmt, field_no);
+	lua_createtable(L, 0, 2);
+	lua_pushstring(L, fmt->dict->names[field_no]);
+	lua_setfield(L, -2, "name");
+	lua_pushstring(L, field_type_strs[field->type]);
+	lua_setfield(L, -2, "type");
+}
+
+/*
+ * Next function of `{i}pairs`.
+ */
+static int
+luaT_tuple_format_iterator_next(struct lua_State *L)
+{
+	if (lua_isnil(L, 1)) {
+		lua_pushnil(L);
+		return 1;
+	}
+	struct tuple_format *fmt = lbox_check_tuple_format(L, 1);
+	uint32_t field_cnt = fmt->dict->name_count;
+	if (field_cnt == 0) {
+		lua_pushnil(L);
+		return 1;
+	}
+
+	if (lua_isnil(L, 2)) {
+		lua_pushnumber(L, 1);
+		push_tuple_field_format(L, fmt, 0);
+		return 2;
+	}
+
+	uint32_t i = luaL_touint64(L, 2);
+	if (i >= field_cnt) {
+		lua_pushnil(L);
+		return 1;
+	}
+	lua_pushnumber(L, i + 1);
+	push_tuple_field_format(L, fmt, i);
+	return 2;
+}
+
+/*
+ * Returns (i, {name, type}) pairs.
+ */
+static int
+lbox_tuple_format_pairs(struct lua_State *L)
+{
+	lbox_check_tuple_format(L, 1);
+	lua_rawgeti(L, LUA_REGISTRYINDEX, luaT_tuple_format_iterator_next_ref);
+	lua_insert(L, 1);
+	lua_pushnil(L);
+	return 3;
+}
+
+/*
+ * Returns an array tuple field formats.
+ */
+static int
+lbox_tuple_format_serialize(struct lua_State *L)
+{
+	struct tuple_format *fmt = lbox_check_tuple_format(L, 1);
+	uint32_t field_cnt = fmt->dict->name_count;
+	lua_createtable(L, (int)field_cnt, 0);
+	for (uint32_t i = 0; i < field_cnt; ++i) {
+		push_tuple_field_format(L, fmt, i);
+		lua_rawseti(L, -2, (int)i + 1);
+	}
+	return 1;
+}
+
+/*
+ * Creates a cdata object and pushes format onto Lua stack.
+ */
+static int
+lbox_push_tuple_format(struct lua_State *L, struct tuple_format *format)
+{
+	struct tuple_format **ptr = (struct tuple_format **)
+		luaL_pushcdata(L, CTID_STRUCT_TUPLE_FORMAT_PTR);
+	*ptr = format;
+	tuple_format_ref(format);
+	return 1;
+}
+
+static int
+lbox_tuple_format_new(struct lua_State *L)
+{
+	assert(CTID_STRUCT_TUPLE_FORMAT_PTR != 0);
+	int top = lua_gettop(L);
+	if (top == 0)
+		return lbox_push_tuple_format(L, tuple_format_runtime);
+	assert(top == 2 && lua_istable(L, 1) && lua_isboolean(L, 2));
+	uint32_t count = lua_objlen(L, 1);
+	if (count == 0)
+		return lbox_push_tuple_format(L, tuple_format_runtime);
+	bool space_fields_needed = lua_toboolean(L, 2);
+	size_t size;
+	struct region *region = &fiber()->gc;
+	size_t region_svp = region_used(region);
+	struct field_def *fields =
+		(struct field_def *)xregion_alloc_array(region,
+							struct field_def, count,
+							&size);
+	for (uint32_t i = 0; i < count; ++i) {
+		size_t len;
+		fields[i] = field_def_default;
+		lua_pushinteger(L, i + 1);
+		lua_gettable(L, 1);
+
+		lua_pushstring(L, "name");
+		lua_gettable(L, -2);
+		assert(!lua_isnil(L, -1));
+		const char *name = lua_tolstring(L, -1, &len);
```

Comment:
`xregion_alloc_array()` above will leak if `lua_tolstring()` throws. Same about other region allocs and `lua_tolstring()` calls. This is easy to achieve if you make an object with a `__tostring` meta method which does something non-trivial (or trivially calls `error()` in certain cases).

Reply:
Actually, this is an internal method, and and sanely we will only pass normal strings.

BTW, the [original](https://github.com/tarantool/tarantool/blob/22d7636b521fd2e18104852ba1e0c206825f92a5/src/box/lua/misc.cc#L415) implementation didn't deal with this.

Reply:
Oh, if it was bad before, then sure lets leave it bad